cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/upgrade/TemplateSetUpgrader.java

```diff
@@ -0,0 +1,312 @@
+
+/**
+ * Upgrader for the templates from v2_1 to v3_0 that splits the monolithic
+ * template structure into version 3.0 with template sets
+ */
+public class TemplateSetUpgrader {
+
+	/** Logger instance. */
+	private static final Logger LOG = LoggerFactory.getLogger(TemplateSetUpgrader.class);
+	/** Mapper factory instance. */
+	private MapperFactory mapperFactory;
+	/** Mapper Facade instance. */
+	private MapperFacade mapper;
+
+	/**
+	 * Creates a new {@link TemplateSetUpgrader} instance to upgrade v2.1 ContextConfigurations to v3.0
+	 */
+	public TemplateSetUpgrader() {
+
+		this.mapperFactory = new DefaultMapperFactory.Builder().useAutoMapping(true).mapNulls(true).build();
+		this.mapperFactory
+				.classMap(com.devonfw.cobigen.impl.config.entity.io.ContainerMatcher.class,
+						com.devonfw.cobigen.impl.config.entity.io.v3_0.ContainerMatcher.class)
+				.field("retrieveObjectsRecursively:{isRetrieveObjectsRecursively|setRetrieveObjectsRecursively(new Boolean(%s))|type=java.lang.Boolean}",
+						"retrieveObjectsRecursively:{isRetrieveObjectsRecursively|setRetrieveObjectsRecursively(new Boolean(%s))|type=java.lang.Boolean}")
+				.byDefault().register();
+		this.mapperFactory.classMap(com.devonfw.cobigen.impl.config.entity.io.Trigger.class,
+				com.devonfw.cobigen.impl.config.entity.io.v3_0.Trigger.class).byDefault().register();
+		this.mapperFactory.classMap(com.devonfw.cobigen.impl.config.entity.io.Matcher.class,
+				com.devonfw.cobigen.impl.config.entity.io.v3_0.Matcher.class).byDefault().register();
+		this.mapperFactory
+				.classMap(com.devonfw.cobigen.impl.config.entity.io.ContainerMatcher.class,
+						com.devonfw.cobigen.impl.config.entity.io.v3_0.ContainerMatcher.class)
+				.field("retrieveObjectsRecursively:{isRetrieveObjectsRecursively|setRetrieveObjectsRecursively(new Boolean(%s))|type=java.lang.Boolean}",
+						"retrieveObjectsRecursively:{isRetrieveObjectsRecursively|setRetrieveObjectsRecursively(new Boolean(%s))|type=java.lang.Boolean}")
+				.byDefault().register();
+		this.mapper = mapperFactory.getMapperFacade();
+	}
+
+
+	/**
+	 * Upgrades the ContextConfiguration from v2.1 to the new structure from v3.0. The
+	 * monolithic pom and context files will be split into multiple files
+	 * corresponding to every template set that will be created.
+	 *
+	 * @param {@link Path} Path to the context.xml that will be upgraded
+	 * @return {@link Map} collection that contains the upgraded v3.0 {@link com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration}
+	 * 			as key and a {@link Path} for the new location of the context.xml as value
+	 * @throws Exception
+	 */
+	public Map<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration, Path> upgradeTemplatesToTemplateSets(
+			Path contextLocation) throws Exception {
+		Path context = analyseStructure(contextLocation);
+		ContextConfiguration contextConfiguration = getContextConfiguration(context);
+
+		Path cobigenDir = context;
```

Comment:
I really don't understand why you need this variable. Better rename context to cobiGenDir instead of redefining another variable. Especially if context is not used in this method apart from a new definition.